In [1]:
import pandas as pd
import numpy as np

In [2]:
school_data_csv = "Resources/schools_complete.csv"
student_data_csv = "Resources/students_complete.csv"

In [3]:
school_data = pd.read_csv(school_data_csv)
school_data.head(10)

School ID            school_name      type  size   budget
0          0      Huang High School  District  2917  1910635
1          1   Figueroa High School  District  2949  1884411
2          2    Shelton High School   Charter  1761  1056600
3          3  Hernandez High School  District  4635  3022020
4          4    Griffin High School   Charter  1468   917500
5          5     Wilson High School   Charter  2283  1319574
6          6    Cabrera High School   Charter  1858  1081356
7          7     Bailey High School  District  4976  3124928
8          8     Holden High School   Charter   427   248087
9          9       Pena High School   Charter   962   585858

In [4]:
student_data = pd.read_csv(student_data_csv)
student_data.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  
0             66          79  
1             94          61  
2             90          60  
3             67          58  
4             97          84

In [5]:
# Merge the data frames
school_data_complete = pd.merge(student_data, school_data, how="left", on=["school_name", "school_name"])
school_data_complete.head(10)

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   
5           5      Bryan Miranda      M   9th  Huang High School   
6           6      Sheena Carter      F  11th  Huang High School   
7           7       Nicole Baker      F  12th  Huang High School   
8           8       Michael Roth      M  10th  Huang High School   
9           9     Matthew Greene      M  10th  Huang High School   

   reading_score  math_score  School ID      type  size   budget  
0             66          79          0  District  2917  1910635  
1             94          61          0  District  2917  1910635  
2             90          60          0  District  2917  1910635  
3             67          58          0  District  2917  1910635  
4             97          84          0  District  2917  1910635  
5             94          94          0  District  2917  1910635  
6             82          80          0  District  2917  1910635  
7             96          69          0  District  2917  1910635  
8             95          87          0  District  2917  1910635  
9             96          84          0  District  2917  1910635

# DISTRICT SUMMARY

In [6]:
# Count the number of unique schools 
unique_df = school_data_complete["school_name"].unique()
unique_df

array(['Huang High School', 'Figueroa High School', 'Shelton High School',
       'Hernandez High School', 'Griffin High School',
       'Wilson High School', 'Cabrera High School', 'Bailey High School',
       'Holden High School', 'Pena High School', 'Wright High School',
       'Rodriguez High School', 'Johnson High School', 'Ford High School',
       'Thomas High School'], dtype=object)

In [7]:
length_df = len(school_data_complete["school_name"].unique())
length_df

15

In [8]:
# Count the number of students 
students = len(school_data_complete["Student ID"].unique())
students

39170

In [9]:
# Total Budget 
budget = (school_data_complete["budget"].unique().sum())
budget

24649428

In [10]:
#Average math score 
math_average = school_data_complete["math_score"].mean()
math_average

78.98537145774827

In [11]:
#Average reading score 
reading_average = school_data_complete["reading_score"].mean()
reading_average

81.87784018381414

In [12]:
# Percentage of students who passed math (math scores greather than or equal to 70)
passing_math_count = school_data_complete[(school_data_complete["math_score"] >= 70)].count()["student_name"]
passing_math_percentage = passing_math_count / float(students) * 100
passing_math_percentage

74.9808526933878

In [13]:
# Percentage of studenst who passed reading (score greater than or equal to 70)
passing_reading_count = school_data_complete[(school_data_complete["reading_score"] >= 70)].count()["student_name"]
passing_reading_percentage = passing_reading_count / float(students) * 100
passing_reading_percentage

85.80546336482001

In [14]:
# Percentage passing math and reading 
passing_math_reading_count = school_data_complete[
    (school_data_complete["math_score"] >= 70) & (school_data_complete["reading_score"] >= 70)
].count()["student_name"]
overall_passing_rate = passing_math_reading_count /  float(students) * 100
overall_passing_rate

65.17232575950983

In [15]:
# Create a high-level snapshot of the district's key metrics in a DataFrame
district_summary = pd.DataFrame({"Total Schools": [length_df], "Total Students": [students],
                                 "Total Budget": [budget], "Average Math Score": [math_average],
                                 "Average Reading Score": [reading_average], "% Passing Math": [passing_math_percentage],
                                 "% Passing Reading": [passing_reading_percentage], 
                                 "% Passing Overall": [overall_passing_rate]})
                                 
# Formatting
district_summary["Total Students"] = district_summary["Total Students"].map("{:,}".format)
district_summary["Total Budget"] = district_summary["Total Budget"].map("${:,.2f}".format)

# Display the DataFrame
district_summary

Total Schools Total Students    Total Budget  Average Math Score  \
0             15         39,170  $24,649,428.00           78.985371   

   Average Reading Score  % Passing Math  % Passing Reading  % Passing Overall  
0               81.87784       74.980853          85.805463          65.172326

WRITTEN ANALYSIS - DISTRICT SUMMARY

There are 15 total schools in this district, with a total of 39,170 students. The total district budget (sum of all school budgets) is $246,494,28. The average math score is approximately 79, with an average of 75% of students passing math with a 70 or higher. The average reading score is approximately 82, with an average of 86% of students passing reading with a 70 or higher. The overall percentage of students passing math and reading is approximately 65%.

# SCHOOL SUMMARY

In [16]:
# Use the code provided to select the school type
school_types = school_data.set_index(["school_name"])["type"]
school_types

school_name
Huang High School        District
Figueroa High School     District
Shelton High School       Charter
Hernandez High School    District
Griffin High School       Charter
Wilson High School        Charter
Cabrera High School       Charter
Bailey High School       District
Holden High School        Charter
Pena High School          Charter
Wright High School        Charter
Rodriguez High School    District
Johnson High School      District
Ford High School         District
Thomas High School        Charter
Name: type, dtype: object

In [17]:
# Calculate the total student count
per_school_counts = school_data_complete["school_name"].value_counts()
per_school_counts

Bailey High School       4976
Johnson High School      4761
Hernandez High School    4635
Rodriguez High School    3999
Figueroa High School     2949
Huang High School        2917
Ford High School         2739
Wilson High School       2283
Cabrera High School      1858
Wright High School       1800
Shelton High School      1761
Thomas High School       1635
Griffin High School      1468
Pena High School          962
Holden High School        427
Name: school_name, dtype: int64

In [18]:
# Calculate the total budget for each school 
per_school_budget = school_data_complete.groupby(["school_name"]).mean()["budget"]
per_school_budget

school_name
Bailey High School       3124928.0
Cabrera High School      1081356.0
Figueroa High School     1884411.0
Ford High School         1763916.0
Griffin High School       917500.0
Hernandez High School    3022020.0
Holden High School        248087.0
Huang High School        1910635.0
Johnson High School      3094650.0
Pena High School          585858.0
Rodriguez High School    2547363.0
Shelton High School      1056600.0
Thomas High School       1043130.0
Wilson High School       1319574.0
Wright High School       1049400.0
Name: budget, dtype: float64

In [19]:
# Budget for each student 
per_school_capita = per_school_budget / per_school_counts
per_school_capita

Bailey High School       628.0
Cabrera High School      582.0
Figueroa High School     639.0
Ford High School         644.0
Griffin High School      625.0
Hernandez High School    652.0
Holden High School       581.0
Huang High School        655.0
Johnson High School      650.0
Pena High School         609.0
Rodriguez High School    637.0
Shelton High School      600.0
Thomas High School       638.0
Wilson High School       578.0
Wright High School       583.0
dtype: float64

In [20]:
# Average math and reading score per school. This did not ultimately work since I needed to create a 
# dataframe, but was cool to learn hwo to do 
# per_school_scores = school_data_complete.groupby(["school_name"]).mean(["math_score"])
# per_school_scores.loc[:, ~per_school_scores.columns.isin(['Student ID', 'School ID', 'size', 'budget'])]
#     # https://www.statology.org/pandas-exclude-column/

In [21]:
# Average math scores
per_school_math = school_data_complete.groupby(["school_name"]).mean()["math_score"]
per_school_math

school_name
Bailey High School       77.048432
Cabrera High School      83.061895
Figueroa High School     76.711767
Ford High School         77.102592
Griffin High School      83.351499
Hernandez High School    77.289752
Holden High School       83.803279
Huang High School        76.629414
Johnson High School      77.072464
Pena High School         83.839917
Rodriguez High School    76.842711
Shelton High School      83.359455
Thomas High School       83.418349
Wilson High School       83.274201
Wright High School       83.682222
Name: math_score, dtype: float64

In [22]:
# Average readiing scores
per_school_reading = school_data_complete.groupby(["school_name"]).mean()["reading_score"]
per_school_reading

school_name
Bailey High School       81.033963
Cabrera High School      83.975780
Figueroa High School     81.158020
Ford High School         80.746258
Griffin High School      83.816757
Hernandez High School    80.934412
Holden High School       83.814988
Huang High School        81.182722
Johnson High School      80.966394
Pena High School         84.044699
Rodriguez High School    80.744686
Shelton High School      83.725724
Thomas High School       83.848930
Wilson High School       83.989488
Wright High School       83.955000
Name: reading_score, dtype: float64

In [23]:
# % of students passing math 
#create filtered dataframe from district calcualtions? 
school_passing_math = school_data_complete[(school_data_complete["math_score"] >= 70)]
school_passing_math

Student ID     student_name gender grade         school_name  \
0               0     Paul Bradley      M   9th   Huang High School   
4               4       Bonnie Ray      F   9th   Huang High School   
5               5    Bryan Miranda      M   9th   Huang High School   
6               6    Sheena Carter      F  11th   Huang High School   
8               8     Michael Roth      M  10th   Huang High School   
...           ...              ...    ...   ...                 ...   
39165       39165     Donna Howard      F  12th  Thomas High School   
39166       39166        Dawn Bell      F  10th  Thomas High School   
39167       39167   Rebecca Tanner      F   9th  Thomas High School   
39168       39168     Desiree Kidd      F  10th  Thomas High School   
39169       39169  Carolyn Jackson      F  11th  Thomas High School   

       reading_score  math_score  School ID      type  size   budget  
0                 66          79          0  District  2917  1910635  
4                 97          84          0  District  2917  1910635  
5                 94          94          0  District  2917  1910635  
6                 82          80          0  District  2917  1910635  
8                 95          87          0  District  2917  1910635  
...              ...         ...        ...       ...   ...      ...  
39165             99          90         14   Charter  1635  1043130  
39166             95          70         14   Charter  1635  1043130  
39167             73          84         14   Charter  1635  1043130  
39168             99          90         14   Charter  1635  1043130  
39169             95          75         14   Charter  1635  1043130  

[29370 rows x 11 columns]

In [24]:
# % of students passing reading 
#create filtered dataframe from district calcualtions? 
school_passing_reading = school_data_complete[(school_data_complete["reading_score"] >= 70)]
school_passing_reading

Student ID     student_name gender grade         school_name  \
1               1     Victor Smith      M  12th   Huang High School   
2               2  Kevin Rodriguez      M  12th   Huang High School   
4               4       Bonnie Ray      F   9th   Huang High School   
5               5    Bryan Miranda      M   9th   Huang High School   
6               6    Sheena Carter      F  11th   Huang High School   
...           ...              ...    ...   ...                 ...   
39165       39165     Donna Howard      F  12th  Thomas High School   
39166       39166        Dawn Bell      F  10th  Thomas High School   
39167       39167   Rebecca Tanner      F   9th  Thomas High School   
39168       39168     Desiree Kidd      F  10th  Thomas High School   
39169       39169  Carolyn Jackson      F  11th  Thomas High School   

       reading_score  math_score  School ID      type  size   budget  
1                 94          61          0  District  2917  1910635  
2                 90          60          0  District  2917  1910635  
4                 97          84          0  District  2917  1910635  
5                 94          94          0  District  2917  1910635  
6                 82          80          0  District  2917  1910635  
...              ...         ...        ...       ...   ...      ...  
39165             99          90         14   Charter  1635  1043130  
39166             95          70         14   Charter  1635  1043130  
39167             73          84         14   Charter  1635  1043130  
39168             99          90         14   Charter  1635  1043130  
39169             95          75         14   Charter  1635  1043130  

[33610 rows x 11 columns]

In [25]:
# Schools passing math and reading 
passing_math_and_reading = school_data_complete[
    (school_data_complete["reading_score"] >= 70) & (school_data_complete["math_score"] >= 70)]
passing_math_and_reading

Student ID     student_name gender grade         school_name  \
4               4       Bonnie Ray      F   9th   Huang High School   
5               5    Bryan Miranda      M   9th   Huang High School   
6               6    Sheena Carter      F  11th   Huang High School   
8               8     Michael Roth      M  10th   Huang High School   
9               9   Matthew Greene      M  10th   Huang High School   
...           ...              ...    ...   ...                 ...   
39165       39165     Donna Howard      F  12th  Thomas High School   
39166       39166        Dawn Bell      F  10th  Thomas High School   
39167       39167   Rebecca Tanner      F   9th  Thomas High School   
39168       39168     Desiree Kidd      F  10th  Thomas High School   
39169       39169  Carolyn Jackson      F  11th  Thomas High School   

       reading_score  math_score  School ID      type  size   budget  
4                 97          84          0  District  2917  1910635  
5                 94          94          0  District  2917  1910635  
6                 82          80          0  District  2917  1910635  
8                 95          87          0  District  2917  1910635  
9                 96          84          0  District  2917  1910635  
...              ...         ...        ...       ...   ...      ...  
39165             99          90         14   Charter  1635  1043130  
39166             95          70         14   Charter  1635  1043130  
39167             73          84         14   Charter  1635  1043130  
39168             99          90         14   Charter  1635  1043130  
39169             95          75         14   Charter  1635  1043130  

[25528 rows x 11 columns]

In [43]:
# Calculate the passing rates
per_school_passing_math = school_passing_math.groupby(["school_name"]).count()["student_name"] / per_school_counts * 100
per_school_passing_reading = school_passing_reading.groupby(["school_name"]).count()["student_name"] / per_school_counts * 100
overall_passing_rate = passing_math_and_reading.groupby(["school_name"]).count()["student_name"] / per_school_counts * 100
#per_school_overall_passing_rate = (per_school_passing_math + per_school_passing_reading) / 2

In [45]:
#Create a dataframe
per_school_summary = pd.DataFrame({"School Type": school_types,
                                   "Total Students": per_school_counts,
                                   "Total School Budget": per_school_budget,
                                   "Per Student Budget": per_school_capita,
                                   "Average Math Score": per_school_math,
                                   "Average Reading Score": per_school_reading,
                                   "% Passing Math": per_school_passing_math,
                                   "% Passing Reading": per_school_passing_reading,
                                   "% Overall Passing Rate": overall_passing_rate})
per_school_summary
                                   

School Type  Total Students  Total School Budget  \
Bailey High School       District            4976            3124928.0   
Cabrera High School       Charter            1858            1081356.0   
Figueroa High School     District            2949            1884411.0   
Ford High School         District            2739            1763916.0   
Griffin High School       Charter            1468             917500.0   
Hernandez High School    District            4635            3022020.0   
Holden High School        Charter             427             248087.0   
Huang High School        District            2917            1910635.0   
Johnson High School      District            4761            3094650.0   
Pena High School          Charter             962             585858.0   
Rodriguez High School    District            3999            2547363.0   
Shelton High School       Charter            1761            1056600.0   
Thomas High School        Charter            1635            1043130.0   
Wilson High School        Charter            2283            1319574.0   
Wright High School        Charter            1800            1049400.0   

                       Per Student Budget  Average Math Score  \
Bailey High School                  628.0           77.048432   
Cabrera High School                 582.0           83.061895   
Figueroa High School                639.0           76.711767   
Ford High School                    644.0           77.102592   
Griffin High School                 625.0           83.351499   
Hernandez High School               652.0           77.289752   
Holden High School                  581.0           83.803279   
Huang High School                   655.0           76.629414   
Johnson High School                 650.0           77.072464   
Pena High School                    609.0           83.839917   
Rodriguez High School               637.0           76.842711   
Shelton High School                 600.0           83.359455   
Thomas High School                  638.0           83.418349   
Wilson High School                  578.0           83.274201   
Wright High School                  583.0           83.682222   

                       Average Reading Score  % Passing Math  \
Bailey High School                 81.033963       66.680064   
Cabrera High School                83.975780       94.133477   
Figueroa High School               81.158020       65.988471   
Ford High School                   80.746258       68.309602   
Griffin High School                83.816757       93.392371   
Hernandez High School              80.934412       66.752967   
Holden High School                 83.814988       92.505855   
Huang High School                  81.182722       65.683922   
Johnson High School                80.966394       66.057551   
Pena High School                   84.044699       94.594595   
Rodriguez High School              80.744686       66.366592   
Shelton High School                83.725724       93.867121   
Thomas High School                 83.848930       93.272171   
Wilson High School                 83.989488       93.867718   
Wright High School                 83.955000       93.333333   

                       % Passing Reading  % Overall Passing Rate  
Bailey High School             81.933280               54.642283  
Cabrera High School            97.039828               91.334769  
Figueroa High School           80.739234               53.204476  
Ford High School               79.299014               54.289887  
Griffin High School            97.138965               90.599455  
Hernandez High School          80.862999               53.527508  
Holden High School             96.252927               89.227166  
Huang High School              81.316421               53.513884  
Johnson High School            81.222432               53.539172  
Pena High School               95.945946               90.540541  
Rodriguez High School          80.

In [46]:
# Formatting 
per_school_summary["Total School Budget"] = per_school_summary["Total School Budget"].map("${:,.2f}".format)
per_school_summary["Per Student Budget"] = per_school_summary["Per Student Budget"].map("${:,.2f}".format)

per_school_summary

School Type  Total Students Total School Budget  \
Bailey High School       District            4976       $3,124,928.00   
Cabrera High School       Charter            1858       $1,081,356.00   
Figueroa High School     District            2949       $1,884,411.00   
Ford High School         District            2739       $1,763,916.00   
Griffin High School       Charter            1468         $917,500.00   
Hernandez High School    District            4635       $3,022,020.00   
Holden High School        Charter             427         $248,087.00   
Huang High School        District            2917       $1,910,635.00   
Johnson High School      District            4761       $3,094,650.00   
Pena High School          Charter             962         $585,858.00   
Rodriguez High School    District            3999       $2,547,363.00   
Shelton High School       Charter            1761       $1,056,600.00   
Thomas High School        Charter            1635       $1,043,130.00   
Wilson High School        Charter            2283       $1,319,574.00   
Wright High School        Charter            1800       $1,049,400.00   

                      Per Student Budget  Average Math Score  \
Bailey High School               $628.00           77.048432   
Cabrera High School              $582.00           83.061895   
Figueroa High School             $639.00           76.711767   
Ford High School                 $644.00           77.102592   
Griffin High School              $625.00           83.351499   
Hernandez High School            $652.00           77.289752   
Holden High School               $581.00           83.803279   
Huang High School                $655.00           76.629414   
Johnson High School              $650.00           77.072464   
Pena High School                 $609.00           83.839917   
Rodriguez High School            $637.00           76.842711   
Shelton High School              $600.00           83.359455   
Thomas High School               $638.00           83.418349   
Wilson High School               $578.00           83.274201   
Wright High School               $583.00           83.682222   

                       Average Reading Score  % Passing Math  \
Bailey High School                 81.033963       66.680064   
Cabrera High School                83.975780       94.133477   
Figueroa High School               81.158020       65.988471   
Ford High School                   80.746258       68.309602   
Griffin High School                83.816757       93.392371   
Hernandez High School              80.934412       66.752967   
Holden High School                 83.814988       92.505855   
Huang High School                  81.182722       65.683922   
Johnson High School                80.966394       66.057551   
Pena High School                   84.044699       94.594595   
Rodriguez High School              80.744686       66.366592   
Shelton High School                83.725724       93.867121   
Thomas High School                 83.848930       93.272171   
Wilson High School                 83.989488       93.867718   
Wright High School                 83.955000       93.333333   

                       % Passing Reading  % Overall Passing Rate  
Bailey High School             81.933280               54.642283  
Cabrera High School            97.039828               91.334769  
Figueroa High School           80.739234               53.204476  
Ford High School               79.299014               54.289887  
Griffin High School            97.138965               90.599455  
Hernandez High School          80.862999               53.527508  
Holden High School             96.252927               89.227166  
Huang High School              81.316421               53.513884  
Johnson High School            81.222432               53.539172  
Pena High School               95.945946               90.540541  
Rodriguez High School          80.220055               52.988247  

# HIGHEST-PERFORMING SCHOOLS (BY % OVERALL PASSING)

In [79]:
# Use sort_values function from C:\Users\jessi\OneDrive\Desktop\Class\NU-VIRT-DATA-PT-12-2022-U-LOLC\
#01-Lesson-Plans\04-Data-Analysis-Pandas\3\Activities\02-Stu_Census-Merging\Unsolved\Census.ipynb

highest_schools = per_school_summary.sort_values(["% Overall Passing Rate"], ascending = False)
highest_schools.head(15)

School Type  Total Students Total School Budget  \
Cabrera High School       Charter            1858       $1,081,356.00   
Thomas High School        Charter            1635       $1,043,130.00   
Griffin High School       Charter            1468         $917,500.00   
Wilson High School        Charter            2283       $1,319,574.00   
Pena High School          Charter             962         $585,858.00   
Wright High School        Charter            1800       $1,049,400.00   
Shelton High School       Charter            1761       $1,056,600.00   
Holden High School        Charter             427         $248,087.00   
Bailey High School       District            4976       $3,124,928.00   
Ford High School         District            2739       $1,763,916.00   
Johnson High School      District            4761       $3,094,650.00   
Hernandez High School    District            4635       $3,022,020.00   
Huang High School        District            2917       $1,910,635.00   
Figueroa High School     District            2949       $1,884,411.00   
Rodriguez High School    District            3999       $2,547,363.00   

                      Per Student Budget  Average Math Score  \
Cabrera High School              $582.00           83.061895   
Thomas High School               $638.00           83.418349   
Griffin High School              $625.00           83.351499   
Wilson High School               $578.00           83.274201   
Pena High School                 $609.00           83.839917   
Wright High School               $583.00           83.682222   
Shelton High School              $600.00           83.359455   
Holden High School               $581.00           83.803279   
Bailey High School               $628.00           77.048432   
Ford High School                 $644.00           77.102592   
Johnson High School              $650.00           77.072464   
Hernandez High School            $652.00           77.289752   
Huang High School                $655.00           76.629414   
Figueroa High School             $639.00           76.711767   
Rodriguez High School            $637.00           76.842711   

                       Average Reading Score  % Passing Math  \
Cabrera High School                83.975780       94.133477   
Thomas High School                 83.848930       93.272171   
Griffin High School                83.816757       93.392371   
Wilson High School                 83.989488       93.867718   
Pena High School                   84.044699       94.594595   
Wright High School                 83.955000       93.333333   
Shelton High School                83.725724       93.867121   
Holden High School                 83.814988       92.505855   
Bailey High School                 81.033963       66.680064   
Ford High School                   80.746258       68.309602   
Johnson High School                80.966394       66.057551   
Hernandez High School              80.934412       66.752967   
Huang High School                  81.182722       65.683922   
Figueroa High School               81.158020       65.988471   
Rodriguez High School              80.744686       66.366592   

                       % Passing Reading  % Overall Passing Rate  \
Cabrera High School            97.039828               91.334769   
Thomas High School             97.308869               90.948012   
Griffin High School            97.138965               90.599455   
Wilson High School             96.539641               90.582567   
Pena High School               95.945946               90.540541   
Wright High School             96.611111               90.333333   
Shelton High School            95.854628               89.892107   
Holden High School             96.252927               89.227166   
Bailey High School             81.933280               54.642283   
Ford High School               79.299014               54.289887   
Johnson High School            81.222432               

# LOWEST-PERFORMING SCHOOLS (BY % OVERALL PASSING)

In [77]:
#Switch order to ascending to see bottom performing schools
lowest_schools = per_school_summary.sort_values(["% Overall Passing Rate"], ascending = True)
lowest_schools.head(10)

School Type  Total Students Total School Budget  \
Rodriguez High School    District            3999       $2,547,363.00   
Figueroa High School     District            2949       $1,884,411.00   
Huang High School        District            2917       $1,910,635.00   
Hernandez High School    District            4635       $3,022,020.00   
Johnson High School      District            4761       $3,094,650.00   
Ford High School         District            2739       $1,763,916.00   
Bailey High School       District            4976       $3,124,928.00   
Holden High School        Charter             427         $248,087.00   
Shelton High School       Charter            1761       $1,056,600.00   
Wright High School        Charter            1800       $1,049,400.00   

                      Per Student Budget  Average Math Score  \
Rodriguez High School            $637.00           76.842711   
Figueroa High School             $639.00           76.711767   
Huang High School                $655.00           76.629414   
Hernandez High School            $652.00           77.289752   
Johnson High School              $650.00           77.072464   
Ford High School                 $644.00           77.102592   
Bailey High School               $628.00           77.048432   
Holden High School               $581.00           83.803279   
Shelton High School              $600.00           83.359455   
Wright High School               $583.00           83.682222   

                       Average Reading Score  % Passing Math  \
Rodriguez High School              80.744686       66.366592   
Figueroa High School               81.158020       65.988471   
Huang High School                  81.182722       65.683922   
Hernandez High School              80.934412       66.752967   
Johnson High School                80.966394       66.057551   
Ford High School                   80.746258       68.309602   
Bailey High School                 81.033963       66.680064   
Holden High School                 83.814988       92.505855   
Shelton High School                83.725724       93.867121   
Wright High School                 83.955000       93.333333   

                       % Passing Reading  % Overall Passing Rate  \
Rodriguez High School          80.220055               52.988247   
Figueroa High School           80.739234               53.204476   
Huang High School              81.316421               53.513884   
Hernandez High School          80.862999               53.527508   
Johnson High School            81.222432               53.539172   
Ford High School               79.299014               54.289887   
Bailey High School             81.933280               54.642283   
Holden High School             96.252927               89.227166   
Shelton High School            95.854628               89.892107   
Wright High School             96.611111               90.333333   

                              School Size  
Rodriguez High School   Large (2000-5000)  
Figueroa High School    Large (2000-5000)  
Huang High School       Large (2000-5000)  
Hernandez High School   Large (2000-5000)  
Johnson High School     Large (2000-5000)  
Ford High School        Large (2000-5000)  
Bailey High School      Large (2000-5000)  
Holden High School          Small (<1000)  
Shelton High School    Medium (1000-2000)  
Wright High School     Medium (1000-2000)

ANALYSIS - Overall School Performance 

On average, schools with a higher number of total students (2000-5000) have a much lower percentage of students passing overall compared to schools with a smaller amount of total students (less than 2000). Cabrera High School has the highest percentage in the district of students passing both math and reading (91%), while Rodriguex High School has the lowest (53%). All schools in the district have a higher percentage of students passing reading than of students passing math. 

# MATH SCORES BY GRADE 

In [74]:
# Separate data by grade 
ninth_graders = school_data_complete[(school_data_complete["grade"] == "9th")]
tenth_graders = school_data_complete[(school_data_complete["grade"] == "10th")]
eleventh_graders = school_data_complete[(school_data_complete["grade"] == "11th")]
twelfth_graders = school_data_complete[(school_data_complete["grade"] == "12th")]

In [75]:
# Group by school name and take the mean of each 
grade_nine_scores = ninth_graders.groupby(["school_name"]).mean()["math_score"]
#grade_nine_scores
grade_ten_scores = tenth_graders.groupby(["school_name"]).mean()["math_score"]
#grade_ten_scores
grade_eleven_scores = eleventh_graders.groupby(["school_name"]).mean()["math_score"]
#grade_eleven_scores
grade_twelve_scores = twelfth_graders.groupby(["school_name"]).mean()["math_score"]
#grade_twelve_scores

In [76]:
#Create a dataframe
math_scores_by_grade = pd.DataFrame({"9th Grade": grade_nine_scores, "10th Grade": grade_ten_scores,
                                     "11th Grade": grade_eleven_scores, "12th Grade": grade_twelve_scores})
# Minor data wrangling
math_scores_by_grade.index.name = None
math_scores_by_grade

9th Grade  10th Grade  11th Grade  12th Grade
Bailey High School     77.083676   76.996772   77.515588   76.492218
Cabrera High School    83.094697   83.154506   82.765560   83.277487
Figueroa High School   76.403037   76.539974   76.884344   77.151369
Ford High School       77.361345   77.672316   76.918058   76.179963
Griffin High School    82.044010   84.229064   83.842105   83.356164
Hernandez High School  77.438495   77.337408   77.136029   77.186567
Holden High School     83.787402   83.429825   85.000000   82.855422
Huang High School      77.027251   75.908735   76.446602   77.225641
Johnson High School    77.187857   76.691117   77.491653   76.863248
Pena High School       83.625455   83.372000   84.328125   84.121547
Rodriguez High School  76.859966   76.612500   76.395626   77.690748
Shelton High School    83.420755   82.917411   83.383495   83.778976
Thomas High School     83.590022   83.087886   83.498795   83.497041
Wilson High School     83.085578   83.724422   83.195326   83.035794
Wright High School     83.264706   84.010288   83.836782   83.644986

ANALYSIS - MATH SCORES BY GRADE 

Overall, math scores are fairly consistent across grades in each school, but differ slightly from school to school in the district. Scores range from 76-84% from school to school.   

# READING SCORES BY GRADE 

In [55]:
# Group by school name and take the mean of each 
grade_nine_scores = ninth_graders.groupby(["school_name"]).mean()["reading_score"]
#grade_nine_scores
grade_ten_scores = tenth_graders.groupby(["school_name"]).mean()["reading_score"]
#grade_ten_scores
grade_eleven_scores = eleventh_graders.groupby(["school_name"]).mean()["reading_score"]
#grade_eleven_scores
grade_twelve_scores = twelfth_graders.groupby(["school_name"]).mean()["reading_score"]
#grade_twelve_scores

In [70]:
#Create a dataframe
reading_scores_by_grade = pd.DataFrame({"9th Grade": grade_nine_scores, "10th Grade": grade_ten_scores,
                                     "11th Grade": grade_eleven_scores, "12th Grade": grade_twelve_scores})
# Minor data wrangling
reading_scores_by_grade.index.name = None
reading_scores_by_grade

9th Grade  10th Grade  11th Grade  12th Grade
Bailey High School     81.303155   80.907183   80.945643   80.912451
Cabrera High School    83.676136   84.253219   83.788382   84.287958
Figueroa High School   81.198598   81.408912   80.640339   81.384863
Ford High School       80.632653   81.262712   80.403642   80.662338
Griffin High School    83.369193   83.706897   84.288089   84.013699
Hernandez High School  80.866860   80.660147   81.396140   80.857143
Holden High School     83.677165   83.324561   83.815534   84.698795
Huang High School      81.290284   81.512386   81.417476   80.305983
Johnson High School    81.260714   80.773431   80.616027   81.227564
Pena High School       83.807273   83.612000   84.335938   84.591160
Rodriguez High School  80.993127   80.629808   80.864811   80.376426
Shelton High School    84.122642   83.441964   84.373786   82.781671
Thomas High School     83.728850   84.254157   83.585542   83.831361
Wilson High School     83.939778   84.021452   83.764608   84.317673
Wright High School     83.833333   83.812757   84.156322   84.073171

ANALYSIS - READING SCORES BY GRADE 

Overall, reading scores are fairly consistent across grades, both per grade in school and across each school in the district, 
ranging from approximately 80-84% passing. 

# SCORES BY SCHOOL SPENDING

In [57]:
# Establish the bins 
spending_bins = [0, 585, 630, 645, 680]
labels = ["<$585", "$585-630", "$630-645", "$645-680"]

In [58]:
# Create a copy of the school summary since it has the "Per Student Budget" 
school_spending_df = per_school_summary.copy()

In [59]:
# Use `pd.cut` to categorize spending based on the bins.
school_spending_df["Spending Ranges (Per Student)"] = pd.cut(per_school_capita, spending_bins, labels = labels)
school_spending_df

School Type  Total Students Total School Budget  \
Bailey High School       District            4976       $3,124,928.00   
Cabrera High School       Charter            1858       $1,081,356.00   
Figueroa High School     District            2949       $1,884,411.00   
Ford High School         District            2739       $1,763,916.00   
Griffin High School       Charter            1468         $917,500.00   
Hernandez High School    District            4635       $3,022,020.00   
Holden High School        Charter             427         $248,087.00   
Huang High School        District            2917       $1,910,635.00   
Johnson High School      District            4761       $3,094,650.00   
Pena High School          Charter             962         $585,858.00   
Rodriguez High School    District            3999       $2,547,363.00   
Shelton High School       Charter            1761       $1,056,600.00   
Thomas High School        Charter            1635       $1,043,130.00   
Wilson High School        Charter            2283       $1,319,574.00   
Wright High School        Charter            1800       $1,049,400.00   

                      Per Student Budget  Average Math Score  \
Bailey High School               $628.00           77.048432   
Cabrera High School              $582.00           83.061895   
Figueroa High School             $639.00           76.711767   
Ford High School                 $644.00           77.102592   
Griffin High School              $625.00           83.351499   
Hernandez High School            $652.00           77.289752   
Holden High School               $581.00           83.803279   
Huang High School                $655.00           76.629414   
Johnson High School              $650.00           77.072464   
Pena High School                 $609.00           83.839917   
Rodriguez High School            $637.00           76.842711   
Shelton High School              $600.00           83.359455   
Thomas High School               $638.00           83.418349   
Wilson High School               $578.00           83.274201   
Wright High School               $583.00           83.682222   

                       Average Reading Score  % Passing Math  \
Bailey High School                 81.033963       66.680064   
Cabrera High School                83.975780       94.133477   
Figueroa High School               81.158020       65.988471   
Ford High School                   80.746258       68.309602   
Griffin High School                83.816757       93.392371   
Hernandez High School              80.934412       66.752967   
Holden High School                 83.814988       92.505855   
Huang High School                  81.182722       65.683922   
Johnson High School                80.966394       66.057551   
Pena High School                   84.044699       94.594595   
Rodriguez High School              80.744686       66.366592   
Shelton High School                83.725724       93.867121   
Thomas High School                 83.848930       93.272171   
Wilson High School                 83.989488       93.867718   
Wright High School                 83.955000       93.333333   

                       % Passing Reading  % Overall Passing Rate  \
Bailey High School             81.933280               54.642283   
Cabrera High School            97.039828               91.334769   
Figueroa High School           80.739234               53.204476   
Ford High School               79.299014               54.289887   
Griffin High School            97.138965               90.599455   
Hernandez High School          80.862999               53.527508   
Holden High School             96.252927               89.227166   
Huang High School              81.316421               53.513884   
Johnson High School            81.222432               53.539172   
Pena High School               95.945946               90.540541   
Rodriguez High School          80.220055               

In [60]:
#  Calculate averages for the desired columns. 
spending_math_scores = school_spending_df.groupby(["Spending Ranges (Per Student)"]).mean()["Average Math Score"]
spending_reading_scores = school_spending_df.groupby(["Spending Ranges (Per Student)"]).mean()["Average Reading Score"]
spending_passing_math = school_spending_df.groupby(["Spending Ranges (Per Student)"]).mean()["% Passing Math"]
spending_passing_reading = school_spending_df.groupby(["Spending Ranges (Per Student)"]).mean()["% Passing Reading"]
overall_passing_spending = school_spending_df.groupby(["Spending Ranges (Per Student)"]).mean()["% Overall Passing Rate"]

In [61]:
# Create a dataframe
spending_summary = pd.DataFrame({"Average Math Score": spending_math_scores, 
                                 "Average Reading Score": spending_reading_scores,
                                 "% Passing Math": spending_passing_math,
                                 "% Passing Reading": spending_passing_reading,
                                 "% Overall Passing": overall_passing_spending})
spending_summary

Average Math Score  Average Reading Score  \
Spending Ranges (Per Student)                                              
<$585                                   83.455399              83.933814   
$585-630                                81.899826              83.155286   
$630-645                                78.518855              81.624473   
$645-680                                76.997210              81.027843   

                               % Passing Math  % Passing Reading  \
Spending Ranges (Per Student)                                      
<$585                               93.460096          96.610877   
$585-630                            87.133538          92.718205   
$630-645                            73.484209          84.391793   
$645-680                            66.164813          81.133951   

                               % Overall Passing  
Spending Ranges (Per Student)                     
<$585                                  90.369459  
$585-630                               81.418596  
$630-645                               62.857656  
$645-680                               53.526855

ANALYSIS - SCORES BY SCHOOL SPENDING 

The general trend in the data overall is that the higher the spending range per student, the lower the percentage of students passing both math and reading. Schools that spend less than $585 dollars per student have an overal passing rate of approximately 90%, while schools that spend $645-680 on each student have an overall passing rate of 54%. However, it would be irresponsible to assume that spending less money per student causes higher scores, as there are many other variables that could be responsible for this trend, such as school size, location, grade level, etc.

# SCORES BY SCHOOL SIZE 

In [62]:
# Establish the bins.
size_bins = [0, 1000, 2000, 5000]
labels = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]

In [63]:
# Categorize the spending based on the bins

# Use `pd.cut` on the "Total Students" column of the `per_school_summary` DataFrame.

per_school_summary["School Size"] = pd.cut(per_school_summary["Total Students"], size_bins, labels=labels)
per_school_summary

School Type  Total Students Total School Budget  \
Bailey High School       District            4976       $3,124,928.00   
Cabrera High School       Charter            1858       $1,081,356.00   
Figueroa High School     District            2949       $1,884,411.00   
Ford High School         District            2739       $1,763,916.00   
Griffin High School       Charter            1468         $917,500.00   
Hernandez High School    District            4635       $3,022,020.00   
Holden High School        Charter             427         $248,087.00   
Huang High School        District            2917       $1,910,635.00   
Johnson High School      District            4761       $3,094,650.00   
Pena High School          Charter             962         $585,858.00   
Rodriguez High School    District            3999       $2,547,363.00   
Shelton High School       Charter            1761       $1,056,600.00   
Thomas High School        Charter            1635       $1,043,130.00   
Wilson High School        Charter            2283       $1,319,574.00   
Wright High School        Charter            1800       $1,049,400.00   

                      Per Student Budget  Average Math Score  \
Bailey High School               $628.00           77.048432   
Cabrera High School              $582.00           83.061895   
Figueroa High School             $639.00           76.711767   
Ford High School                 $644.00           77.102592   
Griffin High School              $625.00           83.351499   
Hernandez High School            $652.00           77.289752   
Holden High School               $581.00           83.803279   
Huang High School                $655.00           76.629414   
Johnson High School              $650.00           77.072464   
Pena High School                 $609.00           83.839917   
Rodriguez High School            $637.00           76.842711   
Shelton High School              $600.00           83.359455   
Thomas High School               $638.00           83.418349   
Wilson High School               $578.00           83.274201   
Wright High School               $583.00           83.682222   

                       Average Reading Score  % Passing Math  \
Bailey High School                 81.033963       66.680064   
Cabrera High School                83.975780       94.133477   
Figueroa High School               81.158020       65.988471   
Ford High School                   80.746258       68.309602   
Griffin High School                83.816757       93.392371   
Hernandez High School              80.934412       66.752967   
Holden High School                 83.814988       92.505855   
Huang High School                  81.182722       65.683922   
Johnson High School                80.966394       66.057551   
Pena High School                   84.044699       94.594595   
Rodriguez High School              80.744686       66.366592   
Shelton High School                83.725724       93.867121   
Thomas High School                 83.848930       93.272171   
Wilson High School                 83.989488       93.867718   
Wright High School                 83.955000       93.333333   

                       % Passing Reading  % Overall Passing Rate  \
Bailey High School             81.933280               54.642283   
Cabrera High School            97.039828               91.334769   
Figueroa High School           80.739234               53.204476   
Ford High School               79.299014               54.289887   
Griffin High School            97.138965               90.599455   
Hernandez High School          80.862999               53.527508   
Holden High School             96.252927               89.227166   
Huang High School              81.316421               53.513884   
Johnson High School            81.222432               53.539172   
Pena High School               95.945946               90.540541   
Rodriguez High School          80.220055               

In [64]:
# Calculate averages for the desired columns. 
size_math_scores = per_school_summary.groupby(["School Size"]).mean()["Average Math Score"]
size_reading_scores = per_school_summary.groupby(["School Size"]).mean()["Average Reading Score"]
size_passing_math = per_school_summary.groupby(["School Size"]).mean()["% Passing Math"]
size_passing_reading = per_school_summary.groupby(["School Size"]).mean()["% Passing Reading"]
size_overall_passing = per_school_summary.groupby(["School Size"]).mean()["% Overall Passing Rate"]

In [65]:
size_summary = pd.DataFrame({"Average Math Score": size_math_scores, 
                                 "Average Reading Score": size_reading_scores,
                                 "% Passing Math": size_passing_math,
                                 "% Passing Reading": size_passing_reading,
                                 "% Overall Passing Rate": size_overall_passing})
size_summary

Average Math Score  Average Reading Score  % Passing Math  \
School Size                                                                     
Small (<1000)                83.821598              83.929843       93.550225   
Medium (1000-2000)           83.374684              83.864438       93.599695   
Large (2000-5000)            77.746417              81.344493       69.963361   

                    % Passing Reading  % Overall Passing Rate  
School Size                                                    
Small (<1000)               96.099437               89.883853  
Medium (1000-2000)          96.790680               90.621535  
Large (2000-5000)           82.766634               58.286003

ANALYSIS - SCORES BY SCHOOL SIZE 

On average, it appears that small (1000 or fewer students) to medium (1000-2000 students) size schools have a significantly higher percentage of students passing both math and reading, approximately 90% and 91%, respectively. Large schools (2000-5000 students) have a much lower passing rate overall at approximately 58% of students passing both math and reading. 

# SCORES BY SCHOOL TYPE

In [66]:
# Group the per_school_summary DataFrame by "School Type" and average the results.
type_math_scores = per_school_summary.groupby(["School Type"]).mean()["Average Math Score"]
type_reading_scores = per_school_summary.groupby(["School Type"]).mean()["Average Reading Score"]
type_passing_math = per_school_summary.groupby(["School Type"]).mean()["% Passing Math"]
type_passing_reading = per_school_summary.groupby(["School Type"]).mean()["% Passing Reading"]
type_overall_passing = per_school_summary.groupby(["School Type"]).mean()["% Overall Passing Rate"]

In [67]:
# Use the code provided to select new column data
# average_math_score_by_type = type_math_scores["Average Math Score"]
# average_reading_score_by_type = type_reading_scores["Average Reading Score"]
# average_percent_passing_math_by_type = type_passing_math["% Passing Math"]
# average_percent_passing_reading_by_type = type_passing_reading["% Passing Reading"]
# average_percent_overall_passing_by_type = type_overall_passing["% Overall Passing Rate"]

# I did not end up uing this code, as it kept giving me a syntax error with the variable names that I could not figure out, so I did 
# the following instead to manually find the average overall % passing rate:
avg_passing_rate = (type_passing_math + type_passing_reading) / 2

In [68]:
# Assemble the new data by type into a DataFrame called `type_summary`

type_summary = pd.DataFrame({"Average Math Score": type_math_scores,
                             "Average Reading Score": type_reading_scores,
                             "Average % Passing Math": type_passing_math,
                             "Average % Passing Reading": type_passing_reading,
                             "% Overall Passing Rate": avg_passing_rate})
type_summary            

Average Math Score  Average Reading Score  \
School Type                                              
Charter               83.473852              83.896421   
District              76.956733              80.966636   

             Average % Passing Math  Average % Passing Reading  \
School Type                                                      
Charter                   93.620830                  96.586489   
District                  66.548453                  80.799062   

             % Overall Passing Rate  
School Type                          
Charter                   95.103660  
District                  73.673757

ANALYSIS - SCORES BY SCHOOL TYPE 

On average, charter type schools have a higher percentage overall passing rate than district type schools (95% and 74%, respectively). 